### PR TITLE
Add Search component

### LINKED ISSUE #192: Add FilterBar component
We currently use search bar with filters only in Workloads & Deploy status (2 different components):
image

However, it might be good to take some ideas from Scope search bar and maybe make those 3 converge into one generalized component.
image

In particular:
* Hitting _Enter_ on Deploy/Workloads search bar could wrap the query into a new closable bubble like Scope search currently does
* Adding _Filters_ dropdown to Scope search might take some of the load from the filters in bottom-left corner and at the same time move all filtering into one area of the screen (I think everything but namespace filters could be put in there). Then the default filters could already be shown in search bar like so:
image

cc @foot @bia

### PR DESCRIPTION
**Pull request by jpellizzari**
Closes #192 

Adds a Search component that mimics the one used in scope:
https://drive.google.com/file/d/18inPIxPK40f9bE6t2pAoA1tRxi4sRh4v/view

@fbarl PTAL at the code. @bia Looking for feedback on the design and interaction!

### CODE REVIEW

File: src/components/Search/Search.js [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+import Term from './_SearchTerm';
+
+// Don't make functions async if we are running tests
+const timeout = (cb, t) => {
+  if (process.env.NODE_ENV === 'test') {
+    return cb;
+  }
+
+  return debounce(cb, t);
+};
```
[the remark this answers is not in the document]
Comment (jpellizzari) [reply src/components/Search/Search.js]: This is to keep the tests from having to wait for the timeout to run. They might fail or become indeterministic if we add in a `setTimeout` for `debounce`.

**Comment by bia** [comment]
screen shot 2018-06-22 at 17 25 51
It feels weird (but I guess it's ok) that once I type "blue" the blue item shows up in the list, even though the text is not roundrect-ified, but when I click outside of the text box (even if I've not pressed `enter`) the word `blue` should become a roundrect like`red`.

I like that the background color is bright and highlighted-feeling, but it could be more legible on black. @fbarl and I discussed adding a color like `blue200` for that.

**Comment by jpellizzari** [comment]
@fbarl 

> Left a couple of comments suggestions.. The list of terms in the examples doesn't really get updated for me on my Chrome, does it still work for you?

The example filters things out, so I would not expect anything to show up based on your query. The example is working as intended, but maybe I can used `matchedText` to make it more clear.

**Comment by jpellizzari** [comment]
@fbarl Updated with (most of) the changes you requested. I stopped short of making the component "fully controlled". I would like to keep as much of the state encapsulated in the `Search` component as possible to avoid having to wire up more state logic in the parent or in Redux.

I like the ideas laid out in this blog post by the React team, so I tried to follow them for this component: https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#recommendation-fully-uncontrolled-component-with-a-key

There are now two props `initialQuery` and `initialPinnedTerms` that can be used to initialize the component with data from URL state or wherever.

There are also two handlers; `onChange` the text search change and `onPin` for the pinning/unpinning of terms. The Scope use-case should be covered. The `onChange` handler will also return the terms as the second argument for more simple use cases.

@bia When the input loses focus, the current text value will get turned into a pinned item.

**Comment by jpellizzari** [comment]
Closes https://github.com/weaveworks/ui-components/issues/247

File: src/components/Input/Input.js [hunk 2]
```diff
@@ -90,30 +101,35 @@ class Input extends React.Component {
...
-      'textarea',
-    ]);
+    const {
+      valid,
+      message,
+      label,
+      id,
+      className,
+      textarea,
+      hideValidationMessage,
```
[the remark this answers is not in the document]
Comment (fbarl) [reply src/components/Input/Input.js]: Please? :)

File: src/components/Search/Search.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (fbarl) [reply src/components/Search/Search.js]: > They might fail or become indeterministic if we add in a `setTimeout` for `debounce`.

Fair enough, didn't know that could happen.

**Comment by jpellizzari** [comment]
@fbarl 
> Please? :)

Huh?

**Comment by fbarl** [comment]
> Huh?

Eh, don't know why the comment was detached, it was referring to this previous unaddressed comment:

> Could add this one to `propTypes` (and `defaultProps`).

Not a big deal though, glad you merged! :)

**Pull request merged**